cmd/kubeadm/app/cmd/phases/preflight.go

```diff
@@ -85,3 +119,96 @@ func runPreflightMaster(c workflow.RunData) error {
 
 	return nil
 }
+
+// NewPreflightJoinPhase creates a kubeadm workflow phase that implements preflight checks for a new node join
+func NewPreflightJoinPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:    "preflight",
+		Short:   "Run join pre-flight checks",
+		Long:    "Run join pre-flight checks, functionally equivalent to what is implemented by kubeadm join.",
+		Example: joinPreflightExample,
+		Run:     runPreflightJoin,
+		InheritFlags: []string{
+			options.CfgPath,
+			options.IgnorePreflightErrors,
+			options.TLSBootstrapToken,
+			options.TokenStr,
+			options.ControlPlane,
+			options.APIServerAdvertiseAddress,
+			options.APIServerBindPort,
+			options.NodeCRISocket,
+			options.NodeName,
+			options.FileDiscovery,
+			options.TokenDiscovery,
+			options.TokenDiscoveryCAHash,
+			options.TokenDiscoverySkipCAHash,
+		},
+	}
+}
+
+// runPreflightJoin executes preflight checks logic.
+func runPreflightJoin(c workflow.RunData) error {
+	j, ok := c.(preflightJoinData)
+	if !ok {
+		return errors.New("preflight phase invoked with an invalid data struct")
+	}
+	// Start with general checks
+	klog.V(1).Infoln("[preflight] Running general checks")
+	if err := preflight.RunJoinNodeChecks(utilsexec.New(), j.Cfg(), j.IgnorePreflightErrors()); err != nil {
+		return errors.Wrap(err, "NodeChecks")
+	}
+
+	initCfg, err := j.InitCfg()
+	if err != nil {
+		return errors.Wrap(err, "reading Cluster Configuration")
+	}
+
+	// Continue with more specific checks based on the init configuration
+	klog.V(1).Infoln("[preflight] Running configuration dependant checks")
+	if err := preflight.RunOptionalJoinNodeChecks(utilsexec.New(), initCfg, j.IgnorePreflightErrors()); err != nil {
+		return errors.Wrap(err, "configuration dependant checks")
+	}
+
+	if j.Cfg().ControlPlane != nil {
+		// Checks if the cluster configuration supports
+		// joining a new control plane instance and if all the necessary certificates are provided
+		if err := checkIfReadyForAdditionalControlPlane(initCfg); err != nil {
+			// outputs the not ready for hosting a new control plane instance message
+			ctx := map[string]string{
+				"Error": err.Error(),
+			}
+
+			var msg bytes.Buffer
+			notReadyToJoinControPlaneTemp.Execute(&msg, ctx)
+			return errors.New(msg.String())
+		}
+
+		// run kubeadm init preflight checks for checking all the prequisites
+		fmt.Println("[join] Running pre-flight checks before initializing the new control plane instance")
+		if err := preflight.RunInitMasterChecks(utilsexec.New(), initCfg, j.IgnorePreflightErrors()); err != nil {
+			return errors.Wrap(err, "master checks")
+		}
+
+		fmt.Println("[join] Pulling control-plane images")
```

Comment:
we are sort of alternating between `join` and `preflight` in the output but let's see how it looks like after this review.